## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#6a0d8151-db4a-4686-98bd-d4e05798d961) |
| mint-gateway-azure.sh | :x: [more...](#02b29397-d844-4e3d-aaa3-51d0953d8aba) |
<a name="02b29397-d844-4e3d-aaa3-51d0953d8aba"></a>
### 10251-522639f/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32147
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp db6969b05cdb:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 3 minutes and 32 seconds
(4/15) Running aws-sdk-ruby tests ... done in 17 seconds
(5/15) Running awscli tests ... done in 2 minutes and 47 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 3 minutes and 7 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 34 seconds
(9/15) Running minio-go tests ... FAILED in 5 minutes and 30 seconds
{
  "args": {},
  "duration": 974,
  "error": "The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.",
  "function": "CopyObjectPart(destination, source)",
  "message": "CompleteMultipartUpload call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 6 minutes and 9 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 37 seconds
(11/15) Running minio-py tests ... FAILED in 1 seconds
{
  "name": "minio-py:test_negative_make_bucket_invalid_name",
  "status": "FAIL",
  "args": {
    "location": "default value ('us-east-1')",
    "bucket_name": ".minio-py-test-cf659663-5589-4a57-9ddd-36e6759d563a"
  },
  "message": "NoSuchBucket: message: The specified bucket does not exist.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 147, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 241, in test_negative_make_bucket_invalid_name\n    _call(log_entry, _CLIENT.make_bucket, bucket_name)\n  File \"/mint/run/core/minio-py/tests.py\", line 130, in _call\n    return func(*args, **kwargs)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 335, in make_bucket\n    raise ResponseError(response, method, bucket_name).get_exception()\nminio.error.NoSuchBucket: NoSuchBucket: message: The specified bucket does not exist.\n",
  "function": "make_bucket(bucket_name, location, object_lock)",
  "duration": 409
}
(11/15) Running s3cmd tests ... done in 1 minutes and 23 seconds
(12/15) Running s3select tests ... done in 1 minutes and 4 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="6a0d8151-db4a-4686-98bd-d4e05798d961"></a>
### 10251-522639f/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32021
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4056630a60b3:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 7 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 42 seconds
(8/15) Running minio-dotnet tests ... done in 30 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 20 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-ozgq2qicv4xbs0u2",
      "Object": "srcObject",
      "Encryption": {},
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "source": {
      "Bucket": "minio-go-test-ozgq2qicv4xbs0u2",
      "Object": "srcObject",
      "VersionID": "",
      "MatchETag": "",
      "NoMatchETag": "",
      "MatchModifiedSince": "0001-01-01T00:00:00Z",
      "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
      "MatchRange": false,
      "Start": 0,
      "End": 0,
      "Encryption": null
    }
  },
  "duration": 634,
  "error": "This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes.",
  "function": "CopyObject(destination, source)",
  "message": "CopyObject failed",
  "name": "minio-go: testEncryptedSSES3ToSSES3CopyObject",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 24 seconds
(10/15) Running minio-js tests ... done in 36 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 22 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 2 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/erasure-zones.go
+++ b/cmd/erasure-zones.go
@@ -2088,6 +2088,14 @@ func (z *erasureZones) Health(ctx context.Context, opts HealthOptions) HealthRes
 		}
 	}
 
+	// when maintenance is not specified we don't have
+	// to look at the healing side of the code.
+	if !opts.Maintenance {
+		return HealthResult{
+			Healthy: true,
+		}
+	}
+
 	// check if local disks are being healed, if they are being healed
 	// we need to tell healthy status as 'false' so that this server
 	// is not taken down for maintenance

--- a/pkg/env/web_env.go
+++ b/pkg/env/web_env.go
@@ -29,6 +29,7 @@ import (
 	"net/url"
 	"os"
 	"regexp"
+	"strings"
 	"time"
 
 	"github.com/dgrijalva/jwt-go"
@@ -48,40 +49,16 @@ func RegisterGlobalCAs(CAs *x509.CertPool) {
 	globalRootCAs = CAs
 }
 
-func isValidEnvScheme(scheme string) bool {
-	switch scheme {
-	case webEnvScheme:
-		fallthrough
-	case webEnvSchemeSecure:
-		return true
-	}
-	return false
-}
-
 var (
 	hostKeys = regexp.MustCompile("^(https?://)(.*?):(.*?)@(.*?)$")
 )
 
-func fetchEnvHTTP(envKey string, u *url.URL) (string, error) {
-	switch u.Scheme {
-	case webEnvScheme:
-		u.Scheme = "http"
-	case webEnvSchemeSecure:
-		u.Scheme = "https"
-	}
-
-	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
-	defer cancel()
-
-	var (
-		username, password string
-	)
-
-	envURL := u.String()
+func fetchHTTPConstituentParts(u *url.URL) (username string, password string, envURL string, err error) {
+	envURL = u.String()
 	if hostKeys.MatchString(envURL) {
 		parts := hostKeys.FindStringSubmatch(envURL)
 		if len(parts) != 5 {
-			return "", errors.New("invalid arguments")
+			return "", "", "", errors.New("invalid arguments")
 		}
 		username = parts[2]
 		password = parts[3]
@@ -92,6 +69,31 @@ func fetchEnvHTTP(envKey string, u *url.URL) (string, error) {
 		username = u.User.Username()
 		password, _ = u.User.Password()
 	}
+	return username, password, envURL, nil
+}
+
+func getEnvValueFromHTTP(urlStr, envKey string) (string, error) {
+	u, err := url.Parse(urlStr)
+	if err != nil {
+		return "", err
+	}
+
+	switch u.Scheme {
+	case webEnvScheme:
+		u.Scheme = "http"
+	case webEnvSchemeSecure:
+		u.Scheme = "https"
+	default:
+		return "", errors.New("invalid arguments")
+	}
+
+	username, password, envURL, err := fetchHTTPConstituentParts(u)
+	if err != nil {
+		return "", err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
 
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, envURL+"?key="+envKey, nil)
 	if err != nil {
@@ -161,19 +163,21 @@ func Environ() []string {
 // to fetch ENV values for the env value from a remote server.
 func LookupEnv(key string) (string, bool) {
 	v, ok := os.LookupEnv(key)
-	if ok {
-		u, err := url.Parse(v)
+	if ok && strings.HasPrefix(v, webEnvScheme) {
+		// If env value starts with `env*://`
+		// continue to parse and fetch from remote
+		var err error
+		v, err = getEnvValueFromHTTP(strings.TrimSpace(v), key)
 		if err != nil {
-			return v, true
+			// fallback to cached value if-any.
+			return os.LookupEnv("_" + key)
 		}
-		if !isValidEnvScheme(u.Scheme) {
-			return v, true
-		}
-		v, err = fetchEnvHTTP(key, u)
-		if err != nil {
-			return "", false
-		}
+		// Set the ENV value to _env value,
+		// this value is a fallback in-case of
+		// server restarts when webhook server
+		// is down.
+		os.Setenv("_"+key, v)
 		return v, true
 	}
-	return "", false
+	return v, ok
 }

--- a/pkg/env/web_env_test.go
+++ b/pkg/env/web_env_test.go
@@ -1,0 +1,81 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package env
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/gorilla/mux"
+)
+
+func GetenvHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	if vars["namespace"] != "default" {
+		http.Error(w, "namespace not found", http.StatusNotFound)
+		return
+	}
+	if vars["name"] != "minio" {
+		http.Error(w, "tenant not found", http.StatusNotFound)
+		return
+	}
+	if vars["key"] != "MINIO_ARGS" {
+		http.Error(w, "key not found", http.StatusNotFound)
+		return
+	}
+	w.Write([]byte("http://127.0.0.{1..4}:9000/data{1...4}"))
+	w.(http.Flusher).Flush()
+}
+
+func startTestServer(t *testing.T) *httptest.Server {
+	router := mux.NewRouter().SkipClean(true).UseEncodedPath()
+	router.Methods(http.MethodGet).
+		Path("/webhook/v1/getenv/{namespace}/{name}").
+		HandlerFunc(GetenvHandler).Queries("key", "{key:.*}")
+
+	ts := httptest.NewServer(router)
+	t.Cleanup(func() {
+		ts.Close()
+	})
+
+	return ts
+}
+
+func TestWebEnv(t *testing.T) {
+	ts := startTestServer(t)
+
+	u, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	v, err := getEnvValueFromHTTP(
+		fmt.Sprintf("env://minio:minio123@%s/webhook/v1/getenv/default/minio",
+			u.Host),
+		"MINIO_ARGS")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if v != "http://127.0.0.{1..4}:9000/data{1...4}" {
+		t.Fatalf("Unexpected value %s", v)
+	}
+}

